packages/contracts/contracts/discovery/Batch.sol

```diff
@@ -1,27 +1,161 @@
 // SPDX-License-Identifier: MIT
 pragma solidity =0.8.12;
 
-import {IBatch} from "./interfaces/IBatch.sol";
+import {ERC165Checker} from "@openzeppelin/contracts/utils/introspection/ERC165Checker.sol";
+
 import {ICommon} from "./interfaces/ICommon.sol";
+import {IProject} from "./interfaces/IProject.sol";
+import {IBatch} from "./interfaces/IBatch.sol";
+import {ProjectVoting} from "./ProjectVoting.sol";
+
+import "hardhat/console.sol";
+
+contract Batch is IBatch, ICommon, ProjectVoting {
+    using ERC165Checker for address;
+
+    /// List of addresses for the project in the batch
+    address[] public projects;
 
-contract Batch is IBatch, ICommon {
-    /// address of the projects
-    address[] projects;
     /// number of available slots
-    uint256 slotCount;
-    Period votingPeriod;
+    uint256 public slotCount;
+
+    /// Period for which the batch is open for voting
+    Period public votingPeriod;
+
+    /// Address for the controller contract
+    address public controller;
 
-    /// user => votes
-    mapping(address => uint256) public userVoteCount;
+    /// duration in seconds of each slot
+    uint256 public singleSlotDuration;
 
-    /// projectId => votes
-    mapping(uint256 => uint256) public projectVoteCount;
+    modifier votingPeriodIsSet() {
+        require(votingPeriod.start != 0, "voting period not set");
+        require(votingPeriod.end != 0, "voting period not set");
+        _;
+    }
 
-    constructor(address[] memory _projects, uint256 _slotCount) {
-        require(_projects.length > 0, "_projects.length cannot be 0");
-        require(_slotCount > 0, "_slotCount cannot be 0");
+    modifier inVotingPeriod() {
+        require(
+            votingPeriod.end >= block.timestamp &&
+                votingPeriod.start <= block.timestamp,
+            "outside of voting period"
+        );
+        _;
+    }
 
+    constructor(address[] memory _projects, uint256 _slotCount)
+        ProjectVoting(_projects)
+    {
+        require(_projects.length > 0, "projects must not be empty");
```

Comment:
we should also check that `_slotCount > 0 && _slotCount < _projects.length` (otherwise the batch is kind of meaningless)